cloudstack/data_source_cloudstack_template.go

```diff
@@ -0,0 +1,144 @@
+package cloudstack
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/xanzy/go-cloudstack/cloudstack"
+	"log"
+	"regexp"
+	"time"
+)
+
+func dataSourceCloudstackTemplate() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceCloudstackTemplateRead,
+		Schema: map[string]*schema.Schema{
+			"filter": dataSourceFiltersSchema(),
+			"templatefilter": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			// Computed values
+			"template_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"account": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"created": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"displaytext": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"format": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"hypervisor": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"size": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func dataSourceCloudstackTemplateRead(d *schema.ResourceData, meta interface{}) error {
+	cs := meta.(*cloudstack.CloudStackClient)
+	ltp := cloudstack.ListTemplatesParams{}
+	ltp.SetListall(true)
+	ltp.SetTemplatefilter(d.Get("templatefilter").(string))
+	csTemplates, err := cs.Template.ListTemplates(&ltp)
+	if err != nil {
+		log.Printf("[ERROR] Failed to list templates: %s", err)
+	}
+
+	filters, filtersOk := d.GetOk("filter")
+	var template *cloudstack.Template
+	var templates []*cloudstack.Template
+
+	if filtersOk {
+		for _, t := range csTemplates.Templates {
+			if applyFilters(t, filters.(*schema.Set)) {
+				templates = append(templates, t)
+			}
+		}
+	} else {
+		return fmt.Errorf("No specified filter, too many results.")
+	}
+
+	if len(templates) > 1 {
+		template = mostRecentTemplate(templates)
+	} else if len(templates) == 1 {
+		template = templates[0]
+	} else {
+		return fmt.Errorf("No template is matching with the specified regex.\n")
+	}
+
+	log.Printf("[DEBUG] Selected template: %s\n", template.Displaytext)
+	return templateDescriptionAttributes(d, template)
+}
+
+func templateDescriptionAttributes(d *schema.ResourceData, template *cloudstack.Template) error {
+	d.SetId(template.Id)
```

Comment:
Is this needed for a data source? To set the ID?

Reply:
You have to put an ID on a datasource. I chose to use the template ID like in the [datasource aws_ami](https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/data_source_aws_ami.go#L262).

Reply:
Ah, check... I wasn't sure a datasource also needed an ID. All clear then...